pytorch_lightning/plugins/sharded_plugin.py

```diff
@@ -0,0 +1,86 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import List, Optional, Union
+
+from pytorch_lightning.core.lightning import LightningModule
+from pytorch_lightning.plugins.ddp_plugin import DDPPlugin
+from pytorch_lightning.utilities import rank_zero_only, FAIRSCALE_AVAILABLE
+from pytorch_lightning.utilities.exceptions import MisconfigurationException
+
+if FAIRSCALE_AVAILABLE:
+    from fairscale.optim import OSS
+    from pytorch_lightning.overrides.fairscale import LightningShardedDataParallel
+
+
+class DDPShardedPlugin(DDPPlugin):
+
+    def __init__(self, **kwargs):
+        self._check_fairscale()
+        super().__init__(**kwargs)
+
+    def configure_ddp(
+            self, model: LightningModule, device_ids: List[int]
+    ):
+        self._wrap_optimizers(model)
+        if model.trainer.testing:  # Revert to standard DDP if testing
+            return super().configure_ddp(
+                model=model,
+                device_ids=device_ids
+            )
+        return LightningShardedDataParallel(model, sharded_optimizer=model.trainer.optimizers)
+
+    def optimizer_state(self, optimizer: 'OSS') -> Optional[dict]:
+        optimizer.consolidate_state_dict()
+        return self._optim_state_dict(optimizer)
+
+    def on_before_forward(self, model: LightningModule, *args):
+        return model.transfer_batch_to_device(args, model.trainer.root_gpu)
```

Comment:
it's not merged yet. Waiting for review :] https://github.com/PyTorchLightning/pytorch-lightning/pull/3671

Reply:
I can align the name with the refactor however this hook is separate to the model (exists within the plugin)

Reply:
This is out of scope, since I'll need to modify across accelerators. We can do this after refactors have been merged (trying to keep this to the core feature only)